go/vt/vtgate/planbuilder/jointree.go

```diff
@@ -102,22 +91,62 @@ type (
 		SysTableTableSchema []evalengine.Expr
 		SysTableTableName   map[string]evalengine.Expr
 	}
+)
 
-	joinPlan struct {
-		// columns needed to feed other plans
-		columns []int
-
-		// arguments that need to be copied from the LHS/RHS
-		vars map[string]int
+type (
+	relation interface {
+		tableID() semantics.TableSet
+		tableNames() []string
+	}
 
-		// the children of this plan
-		lhs, rhs joinTree
+	joinTables struct {
+		lhs, rhs relation
+		pred     sqlparser.Expr
+	}
 
-		outer bool
+	routeTable struct {
+		qtable *abstract.QueryTable
+		vtable *vindexes.Table
 	}
 
 	parenTables []relation
 
+	derivedTable struct {
+		// tables contains inner tables that are solved by this plan.
+		// the tables also contain any predicates that only depend on that particular table
+		tables parenTables
+
+		// predicates are the predicates evaluated by this plan
+		predicates []sqlparser.Expr
+
+		// leftJoins are the join conditions evaluated by this plan
+		leftJoins []*outerTable
+
+		alias string
+
+		query *sqlparser.Select
+	}
+)
+
+type (
+	outerTable struct {
+		right relation
```

Comment:
Why did we decide to create 3 type declaration block?

Reply:
I'll add a comment to make it clear. I collected the interface and the structs that implement it in the first two blocks